ext/native/gfx_es2/draw_text_android.cpp

```diff
@@ -0,0 +1,233 @@
+	env_->DeleteLocalRef(jstr);
+	*w = (size >> 16) * fontScaleX_ * dpiScale_;
+	*h = (size & 0xFFFF) * fontScaleY_ * dpiScale_;
+}
+
+void TextDrawerAndroid::DrawString(DrawBuffer &target, const char *str, float x, float y, uint32_t color, int align) {
+	using namespace Draw;
+	std::string text(NormalizeString(std::string(str)));
+	if (text.empty())
+		return;
+	JNIEnv *env;
+	int result = javaVM->GetEnv((void **)&env, JNI_VERSION_1_6);
+	assert(env == env_);
+
+	uint32_t stringHash = hash::Fletcher((const uint8_t *)text.data(), text.size());
+	uint32_t entryHash = stringHash ^ fontHash_ ^ (align << 24);
+
+	target.Flush(true);
+
+	TextStringEntry *entry;
+
+	auto iter = cache_.find(entryHash);
+	if (iter != cache_.end()) {
+		entry = iter->second.get();
+		entry->lastUsedFrame = frameCount_;
+		draw_->BindTexture(0, entry->texture);
+	} else {
+		jstring jstr = env_->NewStringUTF(text.c_str());
+		int len = (int)env_->GetStringUTFLength(jstr);
+		ILOG("UTF len: %d", len);
+		uint32_t size = env_->CallStaticIntMethod(cls_textRenderer, method_measureText, jstr, curSize_);
+		int imageWidth = (size >> 16);
+		int imageHeight = (size & 0xFFFF);
+		jintArray imageData = (jintArray)env_->CallStaticObjectMethod(cls_textRenderer, method_renderText, jstr, curSize_);
+		env_->DeleteLocalRef(jstr);
+
+		entry = new TextStringEntry();
+		entry->bmWidth = entry->width = imageWidth;
+		entry->bmHeight = entry->height = imageHeight;
+		entry->lastUsedFrame = frameCount_;
+
+		TextureDesc desc{};
+		desc.type = TextureType::LINEAR2D;
+		desc.format = Draw::DataFormat::B4G4R4A4_UNORM_PACK16;
+		desc.width = entry->bmWidth;
+		desc.height = entry->bmHeight;
+		desc.depth = 1;
+		desc.mipLevels = 1;
+
+		uint16_t *bitmapData = new uint16_t[entry->bmWidth * entry->bmHeight];
+		jint* jimage = env_->GetIntArrayElements(imageData, nullptr);
+		int arraySize = env_->GetArrayLength(imageData);
+		assert(arraySize == imageWidth * imageHeight);
+		for (int x = 0; x < entry->bmWidth; x++) {
+			for (int y = 0; y < entry->bmHeight; y++) {
+				uint32_t v = jimage[imageWidth * y + x];
+				v = 0xFFF0 | ((v >> 12) & 0xF);  // Just grab some bits from the green channel.
+				bitmapData[entry->bmWidth * y + x] = (uint16_t)v;
+			}
+		}
+		env_->ReleaseIntArrayElements(imageData, jimage, 0);
+		desc.initData.push_back((uint8_t *)bitmapData);
+		entry->texture = draw_->CreateTexture(desc);
+		delete[] bitmapData;
+		cache_[entryHash] = std::unique_ptr<TextStringEntry>(entry);
+		draw_->BindTexture(0, entry->texture);
+	}
+	float w = entry->bmWidth * fontScaleX_ * dpiScale_;
+	float h = entry->bmHeight * fontScaleY_ * dpiScale_;
+	DrawBuffer::DoAlign(align, &x, &y, &w, &h);
+	target.DrawTexRect(x, y, x + w, y + h, 0.0f, 0.0f, 1.0f, 1.0f, color);
+	target.Flush(true);
+}
+
+void TextDrawerAndroid::ClearCache() {
+	for (auto &iter : cache_) {
+		if (iter.second->texture)
+			iter.second->texture->Release();
+	}
+	cache_.clear();
+	fontMap_.clear();
```

Comment:
Hmm, any reason really to clear this?  If so, should we do in the Windows and Qt versions too?

I guess measuring sizes is not slow?  Most of the time, the same string will get measured multiple times (especially because of text wrapping.)  We cache sizes in the others is why I ask, and clear it here.

-[Unknown]